src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -740,27 +740,31 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
 
     def fromScala2x = unapplyFn.symbol.exists && (unapplyFn.symbol.owner is Scala2x)
 
-    /** Can `subtp` be made to be a subtype of `tp`, possibly by dropping some
-     *  refinements in `tp`?
+    /** Is `subtp` a subtype of `tp` or of some generalization of `tp`?
+     *  The generalizations of a type T are the smallest set G such that
+     *
+     *   - T is in G
+     *   - If a typeref R in G represents a trait, R's superclass is in G.
+     *   - If a type proxy P is not a reference to a class, P's supertype is in G
      */
     def isSubTypeOfParent(subtp: Type, tp: Type)(implicit ctx: Context): Boolean =
       if (subtp <:< tp) true
       else tp match {
-        case tp: RefinedType => isSubTypeOfParent(subtp, tp.parent)
+        case tp: TypeRef if tp.symbol.isClass =>
+          tp.symbol.is(Trait) && isSubTypeOfParent(subtp, tp.parents.head)
+        case tp: TypeProxy => isSubTypeOfParent(subtp, tp.superType)
```

Comment:
Thanks @odersky , I see `RefinedType` is a `TypeProxy`. I mean `RefinedType` referencing a class may be caught by `case tp: TypeProxy`, while the comment says:

> If a type proxy P is **not a reference to a class**, P's supertype is in G

It seems to me the comment needs to be updated to be in accord with the code.